text/0001-dynamic-fees-v1.md

```diff
@@ -0,0 +1,93 @@
+- Feature Name: Dynamic Fees V1 (Configurable Fees)
+- Start Date: (2021-04-20)
+- RFC PR: [mobilecoinfoundation/rfcs#1](https://github.com/mobilecoinfoundation/rfcs/pull/1)
+- MobileCoin Epic: MCC-2301
+
+# Summary
+[summary]: #summary
+
+Decrease the minimum fee to 0.0004 MOB (400uMOB, which is 0.02 USD/tx at a 50 USDMOB exchange rate) per board instruction, allow node operators to override the hard-coded minimum fee, and prevent divergent fees by including fee agreement in the node-to-node attestation / key exchange protocol.
+
+# Motivation
+[motivation]: #motivation
+
+MobileCoin launched with a hard-coded minimum fee of 0.01 MOB. When this constant was set, the value of MobileCoin was assumed to be slightly below 1 USD per coin, resulting in a per-transaction fee of under 0.01 USD (1c). In the intervening 6 months, the value per-coin increased to 50 USD, resulting in a per-transaction fee of 0.50 USD.
+
+This fee is too high to reasonably support the cannonical use-case---purchasing a cup of coffee---and we would like to lower the fee to $0.04 to bring it back in line with that use-case.
+
+However, this creates up a different problem. If the consensus system is roughly capable of completing 50tps (transactions per second), then the cost of conducting a denial of service attack is directly related to the size of the fees. At 0.50 USD/tx, consuming the entire consensus power of the network would cost over 2 MUSD/day.
+
+At the desired 0.04 USD/tx, however, the cost of conducing a denial of service attack would be less than 175 kUSD/day. This may seem like a lot of money, but recall that at peak trading on FTX, 80 MMOB or 4 GUSD changed hands, which gives some sense of scale to those costs (i.e. they are too low).
+
+The second issue is that changing the hard-coded fee requires going through the full release cycle: testnet deployment and testing, mainnet enclave signing, client updates to trust the new enclave, a grace period for client updates, then finally a coordinated mainnet deployment by operators.
+
+In our current environment, it takes at least a month to change the fee, and the fees are simultaneously too high and too low. This must change.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When a node operator starts the consensus service, it may optionally provide a configuration option or command-line argument overriding fee. The fee will be appended (as a string) to the existing to the key exchange "prologue" (a series of bytes that both sides agree upon in advance) when performing node-to-node attestation.
+
+If two nodes with different fees attempt to communicate, the result will be an `AeadError` during the handshake phase.
```

Comment:
Could this error be more descriptive? e.g. `AeadError(FeeMismatch)`

Reply:
Unfortunately no: the noise prologue is just a byte string that each side independently hashes together with the key exchange results to construct the symmetric key, and this proposal is saying "let's extend that to include the fee, as well as the [existing] responder ID", so now, if you get an `AeadError` when you are doing an attestation, it means "you and the node you are connecting do disagree about their responder ID (or there is a MiTM attack that has replaced a pubkey somewhere)", now would be updated to be: "disagree about their responder ID or minimum fee".